ports/stm32/mboot/ui.c

```diff
@@ -0,0 +1,255 @@
+        led_state(LED0, state);
+    }
+}
+
+void led0_update(void) {
+    if (led0_cur_state != LED0_STATE_OFF && systick_ms - led0_ms_interval > 50) {
+        uint8_t rate = (led0_cur_state >> 2) & 0x1f;
+        led0_ms_interval += 50;
+        if (++led0_toggle_count >= rate) {
+            led0_toggle_count = 0;
+        }
+        led_state(LED0, (led0_cur_state & (led0_toggle_count == 0 ? 1 : 2)));
+    }
+}
+
+/******************************************************************************/
+// User button
+
+#if !defined(MBOOT_BOARD_GET_RESET_MODE)
+
+#define RESET_MODE_NUM_STATES (4)
+#define RESET_MODE_TIMEOUT_CYCLES (8)
+#ifdef LED2
+#ifdef LED3
+#define RESET_MODE_LED_STATES 0x8421
+#else
+#define RESET_MODE_LED_STATES 0x7421
+#endif
+#else
+#define RESET_MODE_LED_STATES 0x3210
+#endif
+
+static void usrbtn_init(void) {
+    mp_hal_pin_config(MICROPY_HW_USRSW_PIN, MP_HAL_PIN_MODE_INPUT, MICROPY_HW_USRSW_PULL, 0);
+}
+
+static int usrbtn_state(void) {
+    return mp_hal_pin_read(MICROPY_HW_USRSW_PIN) == MICROPY_HW_USRSW_PRESSED;
+}
+
+int mboot_get_reset_mode(void) {
+    usrbtn_init();
+    int reset_mode = BOARDCTRL_RESET_MODE_NORMAL;
+    if (usrbtn_state()) {
+        // Cycle through reset modes while USR is held
+        // Timeout is roughly 20s, where reset_mode=1
+        systick_init();
+        led_init();
+        reset_mode = 0;
+        for (int i = 0; i < (RESET_MODE_NUM_STATES * RESET_MODE_TIMEOUT_CYCLES + 1) * 32; i++) {
+            if (i % 32 == 0) {
+                if (++reset_mode > RESET_MODE_NUM_STATES) {
+                    reset_mode = BOARDCTRL_RESET_MODE_NORMAL;
+                }
+                uint8_t l = RESET_MODE_LED_STATES >> ((reset_mode - 1) * 4);
+                led_state_all(l);
+            }
+            if (!usrbtn_state()) {
+                break;
+            }
+            mp_hal_delay_ms(19);
+        }
+        // Flash the selected reset mode
+        for (int i = 0; i < 6; i++) {
+            led_state_all(0);
+            mp_hal_delay_ms(50);
+            uint8_t l = RESET_MODE_LED_STATES >> ((reset_mode - 1) * 4);
+            led_state_all(l);
+            mp_hal_delay_ms(50);
+        }
+        mp_hal_delay_ms(300);
+    }
+    return reset_mode;
+}
+
+#endif
+
+/******************************************************************************/
+// State change
+
+#if !defined(MBOOT_BOARD_STATE_CHANGE)
+
+void mboot_state_change(mboot_state_t state, uint32_t arg) {
+    switch (state) {
+        case MBOOT_STATE_DFU_START:
+            led_state_all(0);
+            led0_state(LED0_STATE_SLOW_FLASH);
+            break;
+
+        case MBOOT_STATE_DFU_END:
+            led_state_all(0);
+            break;
+
+        case MBOOT_STATE_FSLOAD_START:
+            break;
+
+        case MBOOT_STATE_FSLOAD_END:
+            // Flash LEDs based on success/failure of update
+            for (int i = 0; i < 4; ++i) {
+                if (arg == 0) {
+                    led_state_all(7);
```

Comment:
While we are touching this code, it could be nice to add an enum or macros for the bit flags to make the intention more obvious. For example, on the line above, this means turn on LED0, LED1 and LED2 but not LED3. But you have to go to the function definition to figure this out.

Reply:
Good idea, done.